File: QuestionAnswering/QA_ICL/backend/fastapi_app/model/qa.py
from __future__ import annotations
from typing import Annotated, Literal, Sequence

from pydantic import BaseModel, Field

from model.nlq2datareq import (
    DataRequest,
    Nlq2DataReqExample,
)
from model.rdf_schema import RDFRelation


class DocumentCollection(BaseModel):
    type: Literal["document_collection"] = "document_collection"
    data: list[dict[str, object]]


class TableDataColumn(BaseModel):
    value: str
    label: str

class TableDataBase(BaseModel):
    columns: list[TableDataColumn]
    data: Sequence[
        dict[str, str | float | Sequence[str] | Sequence[float] | TableDataBase | None]
    ]

    @classmethod
    def from_data(cls, data: Sequence[dict[str, object]]):
        cols: list[TableDataColumn] = []
        cols_set = set()
        for datum in data:
            for k in datum.keys():
                if k not in cols_set:
                    cols.append(TableDataColumn(value=k, label=k))
                    cols_set.add(k)

        for datum in data:
            new_kv: dict[str, cls] = dict()
            for k, v in datum.items():
                if isinstance(v, Sequence) and all(
                    isinstance(elem, dict) for elem in v
                ):
                    new_kv[k] = cls.from_data(v)
            datum.update(new_kv)

        return cls(columns=cols, data=data)


class TableData(TableDataBase):
    type: Literal["table"] = "table"

    @classmethod
    def from_data(cls, data: Sequence[dict[str, object]]):
        return cls(**super().from_data(data).model_dump())


class TypedSeries(BaseModel):
    data: list
    type: str


class ScatterPlotTrace(BaseModel):
    name: str | None = None
    x: TypedSeries
    y: TypedSeries


class ScatterPlotData(BaseModel):
    type: Literal["scatter_plot"] = "scatter_plot"
    title: str | None = None
    traces: list[ScatterPlotTrace] = []


CRS84_URI = "http://www.opengis.net/def/crs/OGC/1.3/CRS84"


class WKTGeometryData(BaseModel):
    type: Literal["wkt_geometry"] = "wkt_geometry"
    srs: Literal["crs84"] = "crs84"
    title: str | None = None
    literal: str


DataItem = Annotated[
    DocumentCollection | TableData | ScatterPlotData | WKTGeometryData,
    Field(discriminator="type"),
]


class QARequestArtifact(BaseModel):
    nlq: str
    data_req: DataRequest
    data: object


class QARequest(BaseModel):
    question: str


class TranslationContext(BaseModel):
    examples: list[Nlq2DataReqExample]
    schema_relations: list[RDFRelation]


class QAResponseMetadata(BaseModel):
    rewritten_question: str | None
    translation_context: TranslationContext
    data_request: DataRequest
    linked_variables: dict[str, list[str]]


class QAResponse(BaseModel):
    request_id: str
    metadata: QAResponseMetadata
    data: list[DataItem]

File: QuestionAnswering/QA_ICL/frontend/next_app_marie/src/lib/model/qa.ts
export interface DocumentCollection {
  type: 'document_collection'
  data: { [key: string]: any }[]
}

export type TableDataValue =
  | undefined
  | null
  | string
  | number
  | string[]
  | number[]
  | TableDataBase

export interface TableDataBase {
  columns: string[]
  data: { [key: string]: TableDataValue }[]
}

export interface TableData extends TableDataBase {
  type: 'table'
}

export interface ChemStructData {
  type: 'xyz' | 'cif'
  iri: string
  label: string
  data: string
}

export interface TypedSeries {
  data: any[]
  type: string
}

export interface ScatterPlotTrace {
  name?: string
  x: TypedSeries
  y: TypedSeries
}

export interface ScatterPlotData {
  type: 'scatter_plot'
  title: string
  traces: ScatterPlotTrace[]
}

export interface WKTGeometryData {
  type: 'wkt_geometry'
  srs: 'crs84'
  title?: string
  literal: string
}

export type DataItem =
  | DocumentCollection
  | TableData
  | ScatterPlotData
  | WKTGeometryData

export interface QARequest {
  question: string
}

export interface SparqlDataReqForm {
  type: 'sparql'
  triplestore: string
  query: string
  pkeys: string[]
}

export interface FuncDataReqForm {
  type: 'func'
  name: string
}

export type DataRequestForm = SparqlDataReqForm | FuncDataReqForm

export interface DataRequest {
  var2cls: { [key: string]: string }
  entity_bindings: { [key: string]: (string | { [key: string]: string })[] }
  const_bindings: { [key: string]: any }
  req_form?: DataRequestForm
  visualise: string[]
}

export interface Nlq2DataReqExample {
  nlq: string
  data_req: DataRequest
}

export interface RDFProperty {
  iri: string
  label?: string
  comment?: string
}

export interface TranslationContext {
  examples: [Nlq2DataReqExample, number][]
  properties: [RDFProperty, number][]
}

export interface QAResponseMetadata {
  rewritten_question?: string
  translation_context: TranslationContext
  data_request: DataRequest
  linked_variables: { [key: string]: string[] }
}

export interface QAResponse {
  request_id: string
  metadata: QAResponseMetadata
  data: DataItem[]
  visualisation: { [key: string]: ChemStructData[] }
}
